Add support for cascading PID

deiger opened the pull request.

deiger commented:
Works really well:
<img width="556" alt="graph" src="https://github.com/ScratMan/HASmartThermostat/assets/45061021/4ec98cbd-55a8-401c-9e19-226682f7b6d6">

deiger commented:
Hey, will you be able to review this PR?

ScratMan reviewed (commented):
I don't get the point to cascade two PIDs in the code. It doubles the PID loops and all the corresponding parameters.

Cascading PIDs is relevant for systems on which you need very accurate regulation while variation on one of the input can create large output changes.
Room temperature regulation criticity level is not so high, I think, and don't require cascaded PID.
Could you please explain how you are using this ?

deiger commented:
> I don't get the point to cascade two PIDs in the code. It doubles the PID loops and all the corresponding parameters.
> 
> Cascading PIDs is relevant for systems on which you need very accurate regulation while variation on one of the input can create large output changes. Room temperature regulation criticity level is not so high, I think, and don't require cascaded PID. Could you please explain how you are using this ?

There is a large delay between the in-floor temperature and the room temperature. If I turn on the heating, the floor temperature can get up to 40&deg;C before the room warms up. The cascading PID helps by allowing capping the max floor temperature, as well as have better regulation, with parameters that properly address the heating delay. Considering that in practice the heating is layered (heating body -> floor, floor -> room), it makes sense for the regulation to be layered as well.

ScratMan commented:
OK, 40°C is the absolute maximum water temperature limit for underfloor heating, that's strange that your system reaches such high temperature.
The system should rather use a heating curve to adjust the water temperature based on outdoor temperature, with gain and offset adjusted to compensate the losses of your house. It would allow to have almost constant heating power delivered to your floor, softening the regulation of the (single) PID.

That's what I'm using at home, using the heating curve on my boiler with it's external temperature sensor ; and a Netatmo thermostat that uses a PID. During cold rainy days, the maximum temperature overshoot is 0.2°C.

deiger commented:
It's an electric system, not water based, so it's just an on/off system.
The mentioned temperature is the temperature measured under the floor - I have two DS18B20 sensors in every room, one under the floor, and the other in the room. The floor tiles take time to heat up, and then radiate the heat.

ScratMan commented:
Wouldn't it be simpler to set two thermostat: one to control the floor temperature when heating, with it's setpoint depending on outside temperature to emulate the heating curve, and the second thermostat (the main one) that will control the first, requesting heating by turning it on and off.
That would be a much simpler and more reliable solution to me compared to trying to maintain a single thermostat with two cascaded PIDs that must still work with only one.

deiger commented:
I'm not sure how that would be easier. I don't really care for the floor temperature itself, besides preventing over-heating, so exposing another thermostat doesn't help.
pidtuner.com works the same ingenerating a pid for either an on/off by pwm, or for temperature mapping. Using the cascading PID makes the thermostat far more accurate in its response, since the intermediary sensor is taken into account.

deiger committed:
Add support for cascading PIDs by controlling an inner climate.

deiger commented:
I've changed the code to have one thermostat controlling another. PTAL.

deiger committed:
Call `set_hvac_mode` on top of `set_temperature`

Apparently even though `climate.set_temperature` accepts an `hvac_mode`, it does not turn the thermostat on.
Also fix `_is_device_active` for climate.

deiger committed:
Only call `set_hvac_mode` if needed

Calling it has adverse effects on the inner thermostat, resulting in rapid turn off and on of the controlled heater.

deiger commented:
Hey, could you please review this?
I've updated the PR 5 months ago...

deiger commented:
Ping again...

ScratMan commented on custom_components/smart_thermostat/climate.py:
```diff
@@ -1008,7 +1014,7 @@ async def _async_heater_turn_off(self, force=False):
                 await self.hass.services.async_call(HA_DOMAIN, service, data)
 
-    async def _async_set_valve_value(self, value: float):
+    async def _async_set_valve_value(self, value: float, hvac_mode: HVACMode):
```
Why adding hvac_mode ? As far as I understand, this is used to switch off the secondary thermostat when the main one is switched OFF, but why not use the async_set_hvac_mode(hvac_mode) to sync the state of both thermostats?

deiger replied:
It's more complex than that:
If `SERVICE_SET_TEMPERATURE` is called when the secondary thermostat is off, it turns it on. This is unlike the case for a light or valve.
Now generally the code currently calls `_async_set_valve_value` on a set of controlled entities. If I moved the if statement for the mode to the caller, I guess we need to move the for loop itself outside, and create a separate method `_async_set_climate_hvac_mode`, and call both on each of the `self.heater_or_cooler_entity`.
Do you prefer that? I tried to reduce the collateral changes in this PR.

ScratMan commented on README.md:
The documentation about cascading is not clear enough. The option being complex to understand, user will need more guidance, with more detailed examples.

deiger replied:
Added.

deiger committed:
Add elaborate details on cascading PID usage.

deiger committed:
Merge branch 'ScratMan:master' into master

deiger commented:
Any update on this?
